### PR TITLE
fix(discord): derive a stable thread key and pin the binding (A2)

### PR DESCRIPTION
**Pull request by chelojimenez**
Second step of the agent-surfaces plan, following #3785 (A1). Discord-only again — `slack-app` and `surface-core` are untouched, so the `deploy-slack-app.yml` trigger on `surface-core/**` stays cold.

## The bug

Thread bindings never matched anything, for two independent reasons:

- `threadId` was `message.id`, so **every message minted a fresh thread key**. A binding keyed on one message could never be found by the next.
- `createThreadBinding` was **never called at all** — nothing was ever written to match.

Net effect: every turn fell through to whoever-spoke-last's default project. Two people in one conversation would silently run against two different projects.

## The derivation

Discord has no single "thread key", so the pair has to be built:

| trigger | `conversationId` | `threadId` |
| --- | --- | --- |
| channel mention | `channelId` | `message.id` |
| inside a thread | `channel.parentId` | `channelId` |

These two rows produce **one key**, which is the whole point. A Discord thread started from a message takes that message's id as its own channel id, and its `parentId` is the channel the mention was posted in — so the binding written by the first turn is exactly the one later in-thread turns look up. This also mirrors Slack, where a top-level message's `threadTs` is its own `ts`: the id of the thread that *would* be created if someone replied.

An orphaned thread (`parentId === null`) falls back to its own id rather than sending `null` through as a channel id.

This lives in a new `discord-app/conversation.js` rather than inline in `app.js`, because `app.js` calls `client.login()` at module scope and can't be imported by a test.

## The binding write

The first user-mode turn now writes the binding and **fails the turn if the write fails**, mirroring `slack-app/listeners/events/run-and-reply.js` and carrying its reasoning across: an unbound thread re-resolves on every reply, so the next person to speak gets *their* default project and creates resources somewhere the initiator never chose. A silent cross-project write is worse than a visible "try again".

Three details worth calling out:

- **A 200 can still be a failed bind.** The backend answers `{created: false, reason: 'project_not_in_org'}` with no `projectId` — nothing was pinned, so it's treated exactly like a thrown error. (This is slightly stricter than Slack's fork, which only fails on a throw. The stricter branch is what the invariant actually calls for.)
- **The loser of a race adopts the winner's project.** Binding creation is first-writer-wins server-side and hands the loser the winning `projectId`, so both racing turns act in the same place instead of one running against its own resolution.
- **The claim is released, not completed.** A bind failure is transient, so a redelivery of the same message should be allowed to run rather than being permanently marked done.

Only `user`-mode turns bind — `legacy` runs on a shared key with no org to attribute the thread to, and the other modes never reach a run.

## Also in this diff

`projectId: process.env.MCPJAM_PROJECT_ID` is gone from the pre-resolution `ref`. Seeding it meant every path that returned before resolution acted in whatever project the operator happened to set — a cross-project write dressed up as a default. The resolved target is now the only source. (The same class of bug in the *approval* handler is bug 4, which A4 fixes; this is just the message path.)

This PR depends on A1 having removed history's `conversationId` filter. `conversationId` is now the parent channel for in-thread turns, and that filter would have matched it against nothing — leaving every threaded turn with an empty history.

## Tests

13 new cases in `discord-app/tests/conversation.test.js` (25 total in the workspace, up from 13).

Identity: channel mention, in-thread message, the **carryover case** (a thread started from a mention derives the mention's key), two messages in one thread collapsing to one key, orphaned-thread fallback, and a client without `isThread`.

Binding: the exact wire body, already-bound threads not rewritten, legacy mode not binding, race-loser adopting the winner's project, a thrown write failing the turn, and a 200-that-pinned-nothing failing the turn.

## Verification

| workspace | result |
| --- | --- |
| `@mcpjam/discord-app` | 25 pass, 0 fail — `tsc --checkJs` and biome clean |
| `@mcpjam/surface-core` | 20 pass, 0 fail (untouched) |
| `@mcpjam/slack-app` | 242 pass, 0 fail (untouched) |

## Not in this PR

A3 (credential split), A4 (approval handler target resolution), A5 (connect-link origin normalization), A6 (hardening), then Phase B. Each lands separately.

One thing I deliberately left alone: the `conversationId` passed to `api.runAgentTurn` is still `message.channelId`, not the derived pair. Changing it would alter server-side conversation grouping — in-thread turns would merge under the parent channel — which is a product decision outside this diff's scope.

---
_Generated by [Claude Code](https://claude.ai/code/session_01M6uPdzFLQtSMBHr4rnEiqd)_

---

> [!NOTE]
> **Medium Risk**
> Changes which MCPJam project Discord threads run against and when turns are allowed to proceed; mistakes could still cause wrong-project writes, but failures are surfaced rather than silent drift.
> 
> **Overview**
> Fixes Discord agent turns that **never matched thread bindings** because every message used `message.id` as `threadId`, and **nothing called** `createThreadBinding`. Turns kept re-resolving to each speaker’s default project.
> 
> Adds **`discord-app/conversation.js`**: channel mentions use `(channelId, message.id)`; in-thread messages use `(parentId, threadId)` so replies share the same key as the mention that started the thread (aligned with Slack’s `threadTs` model). After target resolution, the **first user-mode turn** pins the thread via `ensureThreadBinding`; bind failures **abort the turn**, **release** the dedupe claim (not complete), and warn the user instead of running unbound. Race losers adopt the winner’s `projectId`; empty-success bind responses are treated as failure.
> 
> The mention handler **stops seeding** `ref.projectId` from `MCPJAM_PROJECT_ID` before resolution and uses the derived identity for `resolveTurnTarget`. **13 unit tests** cover identity carryover and binding edge cases.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 590b106a109005965c2d6bd9082030b5d119ee66. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

---
## Summary by cubic
Fixes Discord thread routing by deriving a stable `(conversationId, threadId)` and pinning the thread to the resolved project on the first user turn. Adds safer claim handling to avoid stranded messages, prevents cross‑project writes, and makes bind‑failure logs traceable.

- **Bug Fixes**
  - Stable identity: channel mention → `(conversationId=channelId, threadId=message.id)`; in-thread → `(conversationId=parentId || channelId, threadId=channelId)`. Implemented in `discord-app/conversation.js` with tests.
  - Bind on the first user-mode turn; first-writer-wins, and a race loser adopts the winner’s `projectId`. Skip if already bound, mode ≠ user, or the target lacks `organizationId`/`projectId`.
  - On bind failure (throw or no `projectId`): abort the turn, release the claim before warning, and contain the warning send’s throw to avoid re‑completing the claim. Log the dedupe key and conversation id if the warning cannot be delivered. Do not run unbound.
  - Remove pre-resolution `projectId` seeding from the `ref` to prevent cross‑project actions.

<sup>Written for commit 590b106a109005965c2d6bd9082030b5d119ee66. Summary will update on new commits.</sup>

<a href="https://cubic.dev/pr/MCPJam/inspector/pull/3788?utm_source=github" target="_blank" rel="noopener noreferrer" data-no-image-dialog="true"><picture><source media="(prefers-color-scheme: dark)" srcset="https://www.cubic.dev/buttons/review-in-cubic-dark.svg"><source media="(prefers-color-scheme: light)" srcset="https://www.cubic.dev/buttons/review-in-cubic-light.svg">Review in cubic</picture></a>

### CODE REVIEW

**Comment by chelojimenez** [comment]
### :white_check_mark: **Snyk checks have passed. No issues have been found so far.**

| Status | Scan Engine |  Critical |  High |  Medium |  Low | Total (0) |
|---|---|---|---|---|---|---|
| :white_check_mark: | Open Source Security| 0 | 0 | 0 | 0 | [ 0 issues](https://app.snyk.io/org/marcelo-apk6cgUZZztPZRjPdWkYwH/pr-checks/22189633-a41a-4dae-9f5f-a7f2a45e8608?refs=590b106a109005965c2d6bd9082030b5d119ee66&source=prChecksComment) |

:computer: Catch issues earlier using the plugins for [VS Code](https://marketplace.visualstudio.com/items?itemName=snyk-security.snyk-vulnerability-scanner), [JetBrains IDEs](https://plugins.jetbrains.com/plugin/10972-snyk-security), [Visual Studio](https://marketplace.visualstudio.com/items?itemName=snyk-security.snyk-vulnerability-scanner-vs-2022), and [Eclipse](https://marketplace.eclipse.org/content/snyk-security#external-install-button).

File: discord-app/app.js [hunk 1]
```diff
@@ -193,6 +200,33 @@ client.on(Events.MessageCreate, async (message) => {
...
+				textContent(
+					"I could not pin this thread to a project. Try again in a moment.",
+					"warning",
+				),
+			);
+			// Released, not completed: the bind failure is transient, so a
+			// redelivery of this same message should be allowed to run.
+			if (claims.hasClaimBackend())
+				await claims.releaseEvent(dedupeKey).catch(() => {});
+			return;
```
[the remark this answers is not in the document]
Comment (chelojimenez) [reply discord-app/app.js]: Both parts are valid — fixed in c4c081a. cubic flagged the same ordering independently.

The consequence is worse than "the claim is never released". The outer catch on this handler ends in `claims.completeEvent(dedupeKey, { error: error.message })`, so a rejecting `channel.send` doesn't just skip the release — it actively **completes** the claim. The message would be permanently marked done and could never be redelivered, which is the precise opposite of what this branch exists to do.

Released first, then delivered, with the reason written into the comment so the ordering isn't "tidied" back later. The release failure now logs rather than being swallowed — agreed that it's the one failure that strands a message and it was leaving no trace.

On the two nitpicks:

- **Projectless user-mode coverage** — added. Worth noting it pins a guard rather than a live path: no `resolveTurnTarget` branch returns `user` without a `projectId` (that case is `needs_project`). The test says so, and points at where it would actually be caught if it ever happened — A3 restores the core's `NO_PROJECT` guard in the credential seam, which Discord's custom `getConfig` currently drops.
- **Structured logger** — skipping. There isn't one in `discord-app`; `presence.js`, `delivery.js` and this handler all use `console` directly, so the `logger = console` default already is the house convention. The parameter exists so tests can silence it.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: discord-app/app.js [hunk 2]
```diff
@@ -193,6 +200,42 @@ client.on(Events.MessageCreate, async (message) => {
...
+			// Released, not completed: the bind failure is transient, so a
+			// redelivery of this same message should be allowed to run.
+			//
+			// Release BEFORE delivering. `channel.send` can reject on a rate limit,
+			// a missing permission or a deleted channel, and that throw would escape
+			// to the outer catch — which COMPLETES the claim. The turn would then be
+			// permanently marked done and could never be redelivered, which is the
+			// exact opposite of what this branch is for.
+			if (claims.hasClaimBackend())
+				await claims.releaseEvent(dedupeKey).catch((error) => {
```
[the remark this answers is not in the document]
Comment (chelojimenez) [reply discord-app/app.js:225]: Correct, and my previous fix was incomplete — thanks for re-checking it instead of accepting it. Fixed in d366ab2.

Reordering only guaranteed the release *ran*. It did nothing to stop the outer catch from re-completing the claim afterwards: a rejecting `channel.send` still landed in `completeEvent(dedupeKey, { error })` and stranded the message just the same. The send's rejection is now contained in the branch and logged, so it never reaches that catch. Both properties are required — ordering **and** containment — and the comment now says that instead of claiming ordering alone was sufficient.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: discord-app/app.js [hunk 3]
```diff
@@ -193,6 +200,42 @@ client.on(Events.MessageCreate, async (message) => {
...
+					// Swallowing this hides the one failure that strands the message.
+					console.error(`Could not release claim ${dedupeKey}: ${error}`);
```
[the remark this answers is not in the document]
Comment (chelojimenez) [reply discord-app/app.js:227]: Real mismatch, and I'm deliberately not building a retry or recovery path for it in this PR.

The whole hand-rolled claim block in this handler is scaffolding. B6 of this plan ports Discord from `runTurn` to the core's `runTurnForEvent`, which already owns the full claim lifecycle — claim, release on pre-network failure, complete, and replay of a completed envelope — as one tested unit shared with Slack. A bespoke retry ladder added here would be deleted in that PR, and in the meantime it would be a second implementation of exactly the behavior the consolidation exists to make single. That is the failure mode this plan is organized to avoid.

Two things make the interim acceptable rather than just deferred. The failure is now visible — the release logs with the dedupe key, so it is diagnosable instead of silent, which it wasn't before your first comment. And the blast radius is narrow: Discord's gateway does not redeliver `MESSAGE_CREATE`, so the window this claim protects matters mainly for a process restart mid-turn. That is precisely the case B6 fixes properly, and it is currently broken in a larger way than the TTL — a restart mid-turn delivers nothing at all today.

Leaving this thread open rather than resolving it, so it stays visible as a known gap with a named owner.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: discord-app/app.js [hunk 4]
```diff
@@ -193,6 +200,46 @@ client.on(Events.MessageCreate, async (message) => {
...
+			await delivery
+				.deliver(
+					ref,
+					textContent(
+						"I could not pin this thread to a project. Try again in a moment.",
+						"warning",
+					),
+				)
+				.catch((error) => {
+					console.error(`Could not report the bind failure: ${error}`);
```
[the remark this answers is not in the document]
Comment (chelojimenez) [reply discord-app/app.js]: Took the diagnostic half; declining the fallback half. 590b106.

The premise about the prior behavior doesn't hold. The outer catch's "second delivery" is `delivery.deliver(ref, …)` on the **same channel** that just rejected — same channel, same failure mode. For a missing permission or a deleted channel it fails identically, and when it does the throw escapes the `catch` block itself, so `completeEvent` never runs either and the handler ends in an unhandled rejection. That wasn't a fallback buying the user feedback; it was a duplicate attempt whose only distinct outcome was a worse one.

A best-effort fallback send has the same problem: there is no second channel. If `channel.send` is refusing, no message reaches this user, and another attempt would only re-introduce the completing path this branch exists to avoid. When the surface won't accept a message, "the user gets nothing" is a property of the situation rather than of the error handling — and the code should be honest about that instead of performing a retry that reads like a guarantee.

Rate limits aren't the gap they look like either: discord.js queues and retries internally, so a rejection has already survived that.

What was genuinely missing is the thing you point at — a silently dropped turn with nothing to trace it by. The log now carries the dedupe key and the conversation id, and states why no further delivery is attempted, so the drop is diagnosable instead of invisible.

---
_Generated by [Claude Code](https://claude.ai/code)_
